src/metabase/models/card.clj

```diff
@@ -93,12 +94,17 @@
 (defn populate-query-fields
   "Lift `database_id`, `table_id`, and `query_type` from query definition when inserting/updating a Card."
   [{{query-type :type, :as outer-query} :dataset_query, :as card}]
-  (merge (when-let [{:keys [database-id table-id]} (and query-type
-                                                        (query/query->database-and-table-ids outer-query))]
-           {:database_id database-id
-            :table_id    table-id
-            :query_type  (keyword query-type)})
-         card))
+  (cond->> card
+    ;; mega HACK FIXME -- don't update this stuff when doing deserialization because it might differ from what's in the
+    ;; YAML file and break tests like [[metabase-enterprise.serialization.v2.e2e.yaml-test/e2e-storage-ingestion-test]].
+    ;; The root cause of this issue is that we're generating Cards that have a different Database ID or Table ID from
+    ;; what's actually in their query -- we need to fix [[metabase.test.generate]], but I'm not sure how to do that
+    (not mi/*deserializing?*)
+    (merge (when-let [{:keys [database-id table-id]} (and query-type
+                                                          (query/query->database-and-table-ids outer-query))]
+             {:database_id database-id
+              :table_id    table-id
+              :query_type  (keyword query-type)}))))
```

Comment:
Toucan 2 is optimizes out things that haven't changed (e.g. the `database_id`) before calling `before-update` functions like this, which means `(merge {:database_id ...} card)` actually breaks things in Toucan 2.

This change shouldn't affect anything here since this PR still uses Toucan 1, but we can merge it in now to make sure and then not need to change it in the next PR.